controllers/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -0,0 +1,170 @@
+		controllerNameLong  = fmt.Sprintf("%s/%s/%s", ctx.Namespace, ctx.Name, controllerNameShort)
+	)
+
+	r := NewReconciler(
+		ctx,
+		mgr.GetClient(),
+		ctrl.Log.WithName("controllers").WithName(controlledTypeName),
+		record.New(mgr.GetEventRecorderFor(controllerNameLong)),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(controlledType).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
+		Complete(r)
+}
+
+func NewReconciler(
+	ctx context.Context,
+	client client.Client,
+	logger logr.Logger,
+	recorder record.Recorder) *Reconciler {
+	return &Reconciler{
+		Context:  ctx,
+		Client:   client,
+		Logger:   logger,
+		Recorder: recorder,
+	}
+}
+
+// Reconciler reconciles a VirtualMachineSnapShot object.
+type Reconciler struct {
+	client.Client
+	Context  context.Context
+	Logger   logr.Logger
+	Recorder record.Recorder
+}
+
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinesnapshots,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachinesnapshots/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachines,verbs=get;list;watch;
+// +kubebuilder:rbac:groups=vmoperator.vmware.com,resources=virtualmachines/status,verbs=get;update;patch
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx = pkgcfg.JoinContext(ctx, r.Context)
+
+	vmSnapShot := &vmopv1.VirtualMachineSnapshot{}
+	if err := r.Get(ctx, req.NamespacedName, vmSnapShot); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	vmSnapShotCtx := &pkgctx.VirtualMachineSnapshotContext{
+		Context:                ctx,
+		Logger:                 ctrl.Log.WithName("VirtualMachineSnapShot").WithValues("name", req.Name),
+		VirtualMachineSnapshot: vmSnapShot,
+	}
+
+	patchHelper, err := patch.NewHelper(vmSnapShot, r.Client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to init patch helper for %s: %w", vmSnapShotCtx, err)
+	}
+	defer func() {
+		vmSnapShotCtx.Logger.Info("Patching VirtualMachineSnapShot", "snap", vmSnapShot)
+		if err := patchHelper.Patch(ctx, vmSnapShot); err != nil {
+			if reterr == nil {
+				reterr = err
+			}
+			vmSnapShotCtx.Logger.Error(err, "patch failed")
+		}
+	}()
+
+	if !vmSnapShot.DeletionTimestamp.IsZero() {
+		// Noop.
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ReconcileNormal(vmSnapShotCtx); err != nil {
+		vmSnapShotCtx.Logger.Error(err, "Failed to reconcile VirtualMachineSnapShot")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineSnapshotContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineSnapshot")
+	vmSnapShot := ctx.VirtualMachineSnapshot
+
+	// return early if phase was already set; nothing to do
+	if conditions := ctx.VirtualMachineSnapshot.Status.Conditions; len(conditions) != 0 {
```

Comment:
Don't need the len check here. There is a conditions Get() helper too. Should this be checking that the condition is true?